# SQL vs Python Analysis – Employee Attrition

## Objective
The purpose of this notebook is to validate employee attrition insights
by replicating SQL-style analytical logic using Python (Pandas).

This ensures analytical consistency and demonstrates cross-tool proficiency.


Import Libraries

In [1]:
import pandas as pd

Load Dataset

In [2]:
df = pd.read_csv("../01_dataset/raw/IBM_HR_Analytics_Employee_Attrition_Performance.csv")

In [8]:
df.shape

(1470, 35)

In [3]:
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


## Dataset Overview

The dataset contains 1,470 employee records with 35 features,
including demographics, job role, compensation, and work-life metrics.

## Q1: Attrition by Department

**Business Question:**  
Which departments experience the highest employee attrition?

SQL Logic:
- GROUP BY Department
- COUNT total employees
- COUNT employees where Attrition = 'Yes'
- Calculate Attrition Rate


Python Execution

In [4]:
attrition_by_dept = (
    df.groupby("Department")
      .agg(
          Total_Employees=("Attrition", "count"),
          Attrition_Count=("Attrition", lambda x: (x == "Yes").sum())
      )
)

attrition_by_dept["Attrition_Rate"] = (
    attrition_by_dept["Attrition_Count"] /
    attrition_by_dept["Total_Employees"] * 100
).round(2)

attrition_by_dept.sort_values("Attrition_Rate", ascending=False)


,Total_Employees,Attrition_Count,Attrition_Rate
Department,,,
Sales,446,92,20.63
Human Resources,63,12,19.05
Research & Development,961,133,13.84


**Insight:**  
Sales and Human Resources departments show the highest attrition rates,
indicating higher pressure, target-driven roles, and possible workload imbalance.


## Q2: Attrition by Job Role

In [5]:
jobrole_attrition = (
    df.groupby("JobRole")["Attrition"]
      .value_counts()
      .unstack()
      .fillna(0)
      .sort_values(by="Yes", ascending=False)
)

jobrole_attrition


Attrition,No,Yes
JobRole,,
Laboratory Technician,197,62
Sales Executive,269,57
Research Scientist,245,47
Sales Representative,50,33
Human Resources,40,12
Manufacturing Director,135,10
Healthcare Representative,122,9
Manager,97,5
Research Director,78,2


**Insight:**  
Laboratory Technicians and Sales roles show consistently higher attrition,
suggesting shift stress, performance pressure, and limited growth opportunities.


## Q3: Attrition vs Overtime

In [6]:
overtime_attrition = (
    df.groupby("OverTime")['Attrition']
    .value_counts()
    .unstack()
    .fillna(0)
)

overtime_attrition

Attrition,No,Yes
OverTime,,
No,944,110
Yes,289,127


**Insight:**  
Employees working overtime show significantly higher attrition,
highlighting work-life imbalance as a major exit driver.


## Q4: Monthly Income vs Attrition

In [7]:
df.groupby("Attrition")['MonthlyIncome'].mean().round(2)

Attrition
No     6832.74
Yes    4787.09
Name: MonthlyIncome, dtype: float64

**Insight:**  
Employees who left the organization earn substantially less on average,
confirming compensation as a strong attrition driver.


## Cross-Tool Validation Summary

The results obtained using Python (Pandas) fully align with SQL-based
analysis, confirming data correctness and transformation accuracy.

This step validates the dataset for feature engineering and predictive modeling.
